### PR TITLE
Give response to genCacheKey function

### PR DESCRIPTION
**Pull request by ryanto**
This is so users with custom `getCacheKey` functions can use the same etag fix
from: https://github.com/evanshortiss/express-expeditious/blob/master/lib/cache-key.js#L13

### CODE REVIEW

**Comment by evanshortiss** [comment]
Nice one @ryanto, thanks 😄

**Comment by richardscarrott** [comment]
Hi @evanshortiss, thanks for this package, it's really great!

I'm curious however why one would include the "if-none-match" etag in the cache key as surely this would mean clients who have not yet visited a given page would get a cache miss even though the cached content is valid?

**Comment by evanshortiss** [comment]
@richardscarrott, great question. IIRC, and going on [this comment](https://github.com/evanshortiss/express-expeditious/blob/master/lib/middleware.js#L154-L161) I rationalised it, and now I'm struggling to recall the details exactly so bear with me 😀

TLDR; the module store 2 entries in cache for each route, one with the ETag ("if-none-match") and one without - [this code decides](https://github.com/evanshortiss/express-expeditious/blob/master/lib/middleware.js#L162-L163) which to store based on the response generated by the developer's express route. Using [this code](https://github.com/evanshortiss/express-expeditious/blob/master/lib/middleware.js#L98-L119) the module determines which status code applies for cache hits.

====================================================================

I believe there are 6 scenarios the module needs to cover as illustrated in the table. 

* "if-none-match" row indicates if this header is set on the incoming request
* "cache" row indicates the cache status for the incoming request
* "status" row indicates the status the server will return for the combination of "cache" and "if-none-match"


|  | A  | B  | C  | D | E | F |
|-----|-----|-----|-----|-----|-----|-----|
| if-none-match | no | no | yes | yes | yes | yes |
| cache | hit | miss | miss | miss | hit | hit |
| status | 200 | 200 | 200 | 304 | 200 | 304 |

You might ask, why does it care about a 304 at all. The reason is, if the cache is a miss we forward the request to the developer's route handling code. This code can generate many status codes, but this module only cares about a 200 since it won't cache other responses by default. So, assuming a 200 is generated we store that in cache. Simple...kinda.

Column D is where things were problematic IIRC, and that's why column F exists i.e "if-none-match" being part of the cache key and storing a 304 in cache.

If a client sends a valid "if-none-match" header, but the cache is a miss (maybe it was deleted or expired), we need to use the route handler defined by the developer - this results in a 304 response due to the "if-none-match" being valid. Now, we could choose to not cache this response, but the issue is we end up calling the route handler for every incoming request that has a valid "if-none-match" - not good if we're trying to avoid hitting a slow/expensive route!

So, to solve this we include the "if-none-match"/ETag value in the cache key. This means we can store *two* cache entries for a single endpoint. One for clients who can provide the "if-none-match" and another for clients who cannot, i.e if express generates a 304 vs. 200 - check [the code here](https://github.com/evanshortiss/express-expeditious/blob/master/lib/middleware.js#L162-L163). 

This code means the cache contains something like this for each route:

* `GET-/route' (200 OK version)
* `GET-/route-W/"4c97-lSwvAgbf6V5Uri654ggmHQ"` (304 Not Modified version)

For example if using cURL you might choose *to not pass* the "if-none-match", but you would still expect a fast response from the cache with a 200 response. If you're using cURL and *do pass* the "if-none-match" you also expect a quick response from cache.

Perhaps the implementation needs to be worked on, or simplified. Feel free to pick my brain and poke around the code. I need to think a little more but maybe removing the ETag from the cache key and [keeping this code](https://github.com/evanshortiss/express-expeditious/blob/master/lib/middleware.js#L98-L119) would produce the same result.

**Comment by richardscarrott** [comment]
Thanks for the detailed response @evanshortiss, I understand the problem now.

I wonder if it'd be simpler to just hook into express `res.send` rather than listening to events on the `res` itself, e.g.

```js
app.get('/', (req, res, next) => {
    const originalSend = res.send.bind(res);
    res.send = function (chunk) {
        console.log(this.statusCode); // 200
        if (this.statusCode === 200) {
            // Cache `chunk` against cache key which does not care about `req.headers['if-none-match']`
        }
        // `originalSend` will then determine `req.fresh`ness after 
        originalSend(chunk);
    }
    next();
});

app.get('/', (req, res, next) => {
    res.status(200).send('Hello world');
});
```

That way there wouldn't be the need for a two cache entries for each logical cache key...although the etag would need to be generated even on `if-not-modifed` cache hits.

Also, as a side effect it would potentially fix #54 (although I'm sure #54 could be fixed by working out why `on-finished` isn't handling close correctly).

**Comment by evanshortiss** [comment]
@richardscarrott I had looked into wrapping express `res.send`, but that means this lib loses compatibility with streams. Or the library would also need to wrap `res.pipe`, `res.sendfile`, `res.download` etc. to make sure they can be cached and work with streams.

This is definitely worth revisiting and considering though if things can be simplified and still support pipe/streams/files.

**Comment by richardscarrott** [comment]
Ah, hand't considered streams... it looks like the current implementation buffers a stream in memory, I wonder if it could be directly streamed into redis?

**Comment by richardscarrott** [comment]
Hi @evanshortiss. Just testing out the fix for #54 and updating our `genCacheKey` fn to handle 304s correctly.

Is it correct to say the reason the default `genCacheKey` fn [reads the 'etag' from the `res`](https://github.com/evanshortiss/express-expeditious/blob/conn-close-bug/lib/cache-key.js#L15) when writing to the cache after a 304 is because there's a chance the response returns a 304 due to `if-modified-since` rather than `if-none-match`, meaning the etag wouldn't have made it into the cache key in the [first call](https://github.com/evanshortiss/express-expeditious/blob/conn-close-bug/lib/middleware.js#L70)?

Separately, is it correct to say the inclusion of 'if-none-match' in the cache key could cause the cache to become unnecessarily fragmented, i.e. there'll be many cache keys pointing to the same data, e.g.

```
// Client 1 visited the resource 3 days ago and comes back today
Client1Request: GET / { if-not-modified: 'out-of-date-1' }
Client1Response: { status: 200, etag: 'up-to-date', cache: 'miss' } // Cached under 'out-of-date-1'

// Client 2 visited the resource 2 days ago and comes back today
Client2Request: GET / { if-not-modified: 'out-of-date-2' }
Client2Response: { status: 200, etag: 'up-to-date', cache: 'miss' } // Cached under 'out-of-date-2'

// Client 3 visited the resource 1 day ago and comes back today
Client3Request: GET / { if-not-modified: 'out-of-date-3' }
Client3Response: { status: 200, etag: 'up-to-date', cache: 'miss' } // Cached under 'out-of-date-3'

// etc.

// Client 4 also visited the the resource 1 day ago
Client4Request: GET / { if-not-modified: 'out-of-date-3' }
Client4Response: { status: 200, etag: 'up-to-date', cache: 'hit' } 😄 
```

Basically, I think, if you have a shortish cache expiry and your content frequently changes the cache is going to fragment a lot as you could get many different etags from different clients.

I wonder if it'd make sense to have two functions, namely `getReadCacheKey` and `getWriteCacheKey`:

```
export const getReadCacheKey = (req) => {
    // Include the etag from 'if-none-match' if exists
    const etag = req.headers['if-none-match'];
    return `${genCacheKey(req)${etag ? '-${etag}' : ''}}`;
}

export const getWriteCacheKey = (statusCode, req, res) => {
    // Include the latest etag from the response if 304, otherwise do not include an etag
    return statusCode === 304 ? `${genCacheKey(req)}-${res.get('etag')}` : genCacheKey(req);
}

// util.
const genBaseCacheKey = (req) => `${req.method}-${req.originalUrl}`; // etc.
```

Although there'd still be a lot of misses due to the read cache key potentially coming in with many different etags...the 304 makes this v. challenging.

**Comment by evanshortiss** [comment]
> Ah, I hadn't considered streams... it looks like the current implementation buffers a stream in memory, I wonder if it could be directly streamed directly into redis?

Yep, it will only do it once per request though since locks are used to prevent multiple buffers for the same endpoint.

> is because there's a chance the response returns a 304 due to if-modified-since rather than if-none-match

I didn't test with or read `if-modified-since`, but it could potentially be getting used. Are you seeing anything related to it?

> meaning the etag wouldn't have made it into the cache key in the first call

Yep, that's it. The first call might not have the etag.

> in the cache key could cause the cache to become unnecessarily fragmented, i.e. there'll be many cache keys pointing to the same data, e.g.

> Although there'd still be a lot of misses due to the read cache key potentially coming in with many different etags...the 304 makes this v. challenging.

Yeah it definitely does. I *think* the easiest solution is probably to remove the etag from the key entirely. Then, when a client sends request either with or without `if-none-match` the same cache entry is hit, i.e the 200 OK entry. This [middleware code](https://github.com/evanshortiss/express-expeditious/blob/master/lib/middleware.js#L98-L119) can then determine what response to return, i.e a 304 or 200.

Opened #58 for this. Can look into it over the next few days, but feel free to try it out if you have time 👍
